### DISCUSSION TITLE
Some Way to calculate relation  between diff goroutinues  By channel driven for go

### DISCUSSION THREAD

**Discussion by someview** [open-telemetry/opentelemetry-go#3232] [General]
### Scene  Call In Gorutinues
```
package main

import "time"

type moduleA struct {
}

type moduleB struct {
}

var (
	cha = make(chan int, 10)
	chb = make(chan int, 10)
)

func (a *moduleA) run(cha chan int) {
	for {
		select {
		case num := <-cha:
			// doSomeTask
			chb <- 2 * num
			// doSomeTask
		}
	}
}

func (a *moduleB) run(chb chan int) {
	for {
		select {
		case <-chb:
			// doSomeTask
			time.Sleep(time.Second * 10)
		}
	}
}
func main() {
	exit := make(chan struct{}, 1)
	modA := &moduleA{}
	modB := &moduleB{}
	go modA.run(cha)
	go modB.run(chb)
        cha <- 1   // push msg to queue cha
	<-exit
}
```
 “modB.run” groutinue is driven by channel chb and In "modA.run" goroutinue we put a message to chb.Is there any way to handle this situation when we tracing go program:
1.  In all  chains  how can I obtain modA.run and modB.run relation，they are not connected By context directly
2.  Of course the problem can be solved by making chan's element with parent context information, but It is more invasive for dto transmisted In modules 
3. Sometimes we want to omit some modules In a chain,how can this  be achieved

**Comment by dmathieu** [comment]
Hi,

How about passing a struct to your channel? It can then contain a context.

Something like this (untested):

```golang
type msg struct {
  ctx context.Context
  val int
}

var (
	cha = make(chan msg, 10)
	chb = make(chan msg, 10)
)

func (a *moduleA) run(cha chan int) {
	for {
		select {
		case data := <-cha:
			ctx, span := tracer.Start(data.ctx, "moduleA")
			// doSomeTask
			chb <- msg{ctx, 2 * num}
			// doSomeTask
			span.End()
		}
	}
}

func (a *moduleB) run(chb chan int) {
	for {
		select {
		case data := <-chb:
			ctx, span := tracer.Start(data.ctx, "moduleB")
			// doSomeTask
			time.Sleep(time.Second * 10)
			span.End()
		}
	}
}

func main() {
	exit := make(chan struct{}, 1)
	modA := &moduleA{}
	modB := &moduleB{}
	go modA.run(cha)
	go modB.run(chb)
        cha <- msg{context.Background(), 1}   // push msg to queue cha
	<-exit
}
```

**Reply by someview** [reply]
This way dto is dirty. When module number or dto number increases In a program,It's difficult to manange call realation，workload also increases. Even worse，omiting some call in chains becomes almostly impossible

**Reply by dmathieu** [reply]
The official Go way to pass context around, whether across goroutines or methods is through context.Context.
There is no way to retrieve a context outside of that.

**Reply by someview** [reply]
Now there seems no way to achieve this. optl do not  support method like this:
```
t.Tracer.StartSpan(opName,  parentTraceContext)
// must do something like this：
 t.Tracer.StartSpan(ctx, opName)
 ```
 pass context in channel would introduce variable. Even if we can find some way to do this, eg: using user define producer-consumer queue to replace go channel.we also must pass ctx. I do not think context with  traceInfo is a good idea.especially there is a context type named Context[TraceContext]:generic context.

**Comment by AndrewWinterman** [comment]
HI, I also had this question. Given the stricture against storing contexts on structs (though violated by go stdlib itself), I thought something like the following:

```golang
// Spanned is a message that includes a span context
type Spanned[T any] struct {
	Span trace.SpanContext
	T    T
}
```

then again, it seems like a bit of a weird situation, since my parent span will end after successfully pushing to the channel, as opposed to after the child function call has ended.
